### DISCUSSION TITLE
vcpkg doesn't provide python numpy component

### DISCUSSION THREAD

**Discussion by thomasnal** [microsoft/vcpkg#29645] [Q&A]
**Describe the bug**
vcpkg cannot find python that vcpkg installed.

**Environment**
- OS: Windows
- Compiler: Visual Studio 16 2019
- CMake: 3.25
- Azure build agent

**To Reproduce**
1. cmake -DCMAKE_TOOLCHAIN_FILE=C:/vcpkg/scripts/buildsystems/vcpkg.cmake ` ..
2. Could NOT find Python (missing: Python_NumPy_INCLUDE_DIRS NumPy)

vcpkg.json
```json
{
  "name": "stopt",
  "version": "5.0",
  "dependencies": [
    {
      "name": "python3",
      "version>=": "3.8.3#2"
    },
    "pybind11",
    "boost", 
    "boost-mpi",
    "osi",
    "clp",
    "coinutils",
    "eigen3"
  ],
  "builtin-baseline": "69478c5caafcde4c490bb1fccb960296801dbb5f",

  "overrides": [
    { "name": "python3", "version": "3.8.3#2" }
  ]
}
```
CMakeLists.txt
```cmake
...
  FIND_PACKAGE(Python REQUIRED COMPONENTS Interpreter Development NumPy)
...
```
**Expected behavior**
cmake finds the python installed by vcpkg.

**Failure logs**
```sh
cmake -DCMAKE_TOOLCHAIN_FILE=C:/vcpkg/scripts/buildsystems/vcpkg.cmake ` ..
...
    python3[core]:x64-windows -> 3.8.3#2 -- C:\vcpkg\buildtrees\versioning_\versions\python3\a27ac0326c21362d6a324236c772161ac6130ef5
...
Installing 98/163 python3:x64-windows...
Building python3[core]:x64-windows...
-- Installing port from location: C:\vcpkg\buildtrees\versioning_\versions\python3\a27ac0326c21362d6a324236c772161ac6130ef5
-- Downloading https://github.com/python/cpython/archive/v3.8.3.tar.gz -> python-cpython-v3.8.3.tar.gz...
-- Extracting source C:/vcpkg/downloads/python-cpython-v3.8.3.tar.gz
-- Using source at C:/vcpkg/buildtrees/python3/src/v3.8.3-43b03f2bf0.clean
-- Building C:/vcpkg/buildtrees/python3/src/v3.8.3-43b03f2bf0.clean-Lib-Win/PCBuild/pythoncore.vcxproj for Release
-- Building C:/vcpkg/buildtrees/python3/src/v3.8.3-43b03f2bf0.clean-Lib-Win/PCBuild/pythoncore.vcxproj for Debug
-- Up-to-date: C:/vcpkg/packages/python3_x64-windows/include/python3.8
-- Installing: C:/vcpkg/packages/python3_x64-windows/include/python3.8/abstract.h
...
-- Performing post-build validation
-- Performing post-build validation done
Stored binary cache: "C:\Users\VssAdministrator\AppData\Local\vcpkg\archives\5f\5fcb4db6cb8c6ae049ca834ce75baa49163f3d56638ca5f3f4ca61fac3a8bf99.zip"
Elapsed time to handle python3:x64-windows: 3.186 min
...
CMake Error at C:/Program Files/CMake/share/cmake-3.25/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find Python (missing: Python_NumPy_INCLUDE_DIRS NumPy) (found
  version "2.7.18")
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.25/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  C:/Program Files/CMake/share/cmake-3.25/Modules/FindPython.cmake:561 (find_package_handle_standard_args)
  C:/vcpkg/scripts/buildsystems/vcpkg.cmake:829 (_find_package)
  CMakeLists.txt:216 (FIND_PACKAGE)
-- Configuring incomplete, errors occurred!
See also "D:/a/1/s/build/CMakeFiles/CMakeOutput.log".
See also "D:/a/1/s/build/CMakeFiles/CMakeError.log".
```

**Additional context**
**Alternatives tried**
Set PATH to system installation of Python,
```sh
PATH=/c/Users/VssAdministrator/AppData/Roaming/Python/Python38/Scripts:/c/hostedtoolcache/windows/Python/3.8.10/x64/Scripts:/c/hostedtoolcache/windows/Python/3.8.10/x64
```
Failed in the same way.

**Comment by thomasnal** [comment]
Setting `Python3_ROOT_DIR` according to CMake [docs](https://cmake.org/cmake/help/v3.25/module/FindPython3.html#hints) has no effect,
```sh
          SET Python3_ROOT_DIR=C:/hostedtoolcache/windows/Python/3.8.3/x64
          cmake -DCMAKE_TOOLCHAIN_FILE=C:/vcpkg/scripts/buildsystems/vcpkg.cmake `
                -DPython3_ROOT_DIR=C:/hostedtoolcache/windows/Python/3.8.3/x64 `
                ..
...
CMake Error at C:/Program Files/CMake/share/cmake-3.25/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find Python (missing: Python_NumPy_INCLUDE_DIRS NumPy)
```
Content of `C:/hostedtoolcache/windows/Python/3.8.3/x64`,
```sh
...
C:/hostedtoolcache/windows/Python/3.8.3/x64/libs/python3.lib
C:/hostedtoolcache/windows/Python/3.8.3/x64/libs/python38.lib
C:/hostedtoolcache/windows/Python/3.8.3/x64/libs/_tkinter.lib
C:/hostedtoolcache/windows/Python/3.8.3/x64/LICENSE.txt
C:/hostedtoolcache/windows/Python/3.8.3/x64/NEWS.txt
C:/hostedtoolcache/windows/Python/3.8.3/x64/python-3.8.3-amd64.exe
C:/hostedtoolcache/windows/Python/3.8.3/x64/python.exe
C:/hostedtoolcache/windows/Python/3.8.3/x64/python3.dll
C:/hostedtoolcache/windows/Python/3.8.3/x64/python3.exe
C:/hostedtoolcache/windows/Python/3.8.3/x64/python38.dll
C:/hostedtoolcache/windows/Python/3.8.3/x64/pythonw.exe
...
```
More context before calling cmake,
```sh
>>> python3 -V
Python 3.8.3

>>> pip3 list
Package    Version
---------- -------
numpy      1.23.5
pip        22.3.1
setuptools 41.2.0
```

**Comment by dg0yt** [comment]
I guess vcpkg can find python that python installs. It just doesn't provide the numpy component you ask for.

**Comment by huhuang03** [comment]
Any workaround on this? Manually install numpy and manually find and link it?

**Reply by huhuang03** [reply]
cmake seems use this code to find numpy
```
    execute_process(COMMAND ${${_PYTHON_PREFIX}_INTERPRETER_LAUNCHER} "${_${_PYTHON_PREFIX}_EXECUTABLE}" -c
                            "import sys\ntry: import numpy; sys.stdout.write(numpy.get_include())\nexcept:pass\n"
                    RESULT_VARIABLE _${_PYTHON_PREFIX}_RESULT
                    OUTPUT_VARIABLE _${_PYTHON_PREFIX}_NumPy_PATH
                    ERROR_QUIET
                    OUTPUT_STRIP_TRAILING_WHITESPACE)
```

So first install pip for vcpkg python

```
path_to_vcpkg/installed/tools/python3/python -m ensurepip --upgrade
```

Then use this pip to install numpy
```
path_to_vcpkg/installed/tools/python3/Scripts/pip install numpy
```

This should work.

**Reply by huhuang03** [reply]
Use system python is the prefer way!

**Comment by bennahugo** [comment]
Not familiar with vcpkg or the Windows directory structures, but Just making a note that once installed it should be reasonably simple to detect where the library file is (it should be in the site-packages right next to the library). On a Debian install it is
```ipython
In [1]: import numpy as n; print(n.__version__); print(n.get_include())
1.13.3
/usr/lib/python3/dist-packages/numpy/core/include
```

Cmake can be made to rely on this. E.g. https://github.com/casacore/casacore/blob/master/cmake/FindNUMPY.cmake#LL54C1-L54C67

**Reply by dg0yt** [reply]
IIUC asking python for numpy is what CMake find module does:
https://github.com/Kitware/CMake/blob/ae21371e701ecfb88ec34306b50ebdfe00b26578/Modules/FindPython/Support.cmake#L3742-L3748

Does it need an extra action to have an external python actually provide the include files - maybe not just install runtime packages but also development packages?

**Comment by Roy-Kid** [comment]
Hi, I have another similar problem. I have both vcpkg and conda, but CMake always use vcpkg's python, which can not find numpy in conda environment. I try to use Python3_ROOT_DIR to specify it manually but not work. Do you have any idea with it?

**Comment by ofloveandhate** [comment]
I'd like to get on this train, too.  I'm trying to use vcpkg/Cmake to handle my dependencies on Windows, using the CMake Configure tool.  vcpkg can bring Python, Boost and my other C++ dependencies.  But it won't bring NumPy or my other Python dependencies. So when I try to find Python and require numpy, cmake configure fails.  Boo.

We need a way to say something like "install numpy and these other Python libraries" before CMake starts finding things.  And installing numpy via 

```
path_to_vcpkg/installed/tools/python3/Scripts/pip install numpy
```

feels like a non-starter, because then I have to get my user / environment to do this somehow in the middle, because vcpkg is called by CMake in the Visual Studio Code extension.  I need a way to automate the installation of numpy.

**Reply by ofloveandhate** [reply]
Getting closer, I found the [`vcpkg_get_python_packages`](https://vcpkg.io/en/package/vcpkg-get-python-packages.html) port, but it only supports `native`, so I get this error when I try using it:

```
[cmake] vcpkg-get-python-packages is only supported on 'native', which does not match x64-windows. This usually means that there are known build failures, or runtime problems, when building other platforms. To ignore this and attempt to build vcpkg-get-python-packages anyway, rerun vcpkg with `--allow-unsupported`.
```

**Reply by ofloveandhate** [reply]
Adding

```
"VCPKG_INSTALL_OPTIONS": "--allow-unsupported"
``` 

to my `cacheVariables` in `CMakePresets.json` allowed vcpkg to install `vcpkg-get-python-packages` from `vcpkg.json`.  

But, unfortunately, I just get an error during running CMake that it can't find the command `x_vcpkg_get_python_packages` or `vcpkg_get_python_packages`.  I don't understand.

**Reply by Osyotr** [reply]
You probably need this https://github.com/open-vcpkg/python-registry/

**Reply by dg0yt** [reply]
> Adding
> 
> ```
> "VCPKG_INSTALL_OPTIONS": "--allow-unsupported"
> ```

No... 
It is a script port. It is expected to be installed in the host triplet (= "native"). If used as a dependency, combine it with `"host": true`.

And don't forget to remove the non-host installation.

**Reply by ofloveandhate** [reply]
Thank you.  I think that host-only ports are [documented here](https://learn.microsoft.com/en-us/vcpkg/users/host-dependencies).  These pages have a lot of assumed knowledge baked into them, so I'm struggling a bit to cross the gap to getting it used in my own project.

**Reply by ofloveandhate** [reply]
@dg0yt What I did above, did indeed feel a bit wrong.  Thanks for helping me.  

> installed in the host triplet (= "native"). If used as a dependency, combine it with "host": true

Can you please help me understand the steps to accomplish this, please?  I think I need to modify my `vcpkg.json`, with an entry with `"host": true`.  I now have

```
{
    "dependencies": [
        "gmp",
...
        "boost-python",
        "eigen3",
        "vcpkg-get-python-packages", {
            "name": "vcpkg-get-python-packages",
            "host": true
        }
    ]
}

```

But this doesn't feel right because I have the same name twice.  And it's still

I think I also need to add `vcpkg_add_to_path` to my CMakeLists.txt file, but adding it above my project declaration yields Unknown CMake command, so this isn't right.

**Reply by ofloveandhate** [reply]
> It is expected to be installed in the host triplet (= "native")

Does this mean I need to install it from my command line?  I did this, using the Terminal and `vcpkg install vcpkg-get-python-packages`, and it installed.  But running CMake against my setup now gives the same error as before, 

```
vcpkg-get-python-packages is only supported on 'native', which does not match x64-windows. This usually means that there are known build failures, or runtime problems, when building other platforms. To ignore this and attempt to build vcpkg-get-python-packages anyway, rerun vcpkg with `--allow-unsupported`.
```

Thank you!

**Reply by ofloveandhate** [reply]
Changing to

```
        {
            "name": "vcpkg-get-python-packages",
            "host": true
        }
```
and removing the key for it seems to have helped.  But now I'm back to not being able to call `vcpkg_get_python_packages` from `CMakeLists.txt`; I get Unknown CMake command.  The same thing as when I try to use `vcpkg_add_to_path`, to try to help CMake be aware of the new command.  I'm missing something to help CMake be aware of the vcpkg stuff.

**Reply by dg0yt** [reply]
The first thing is easy.
~~~diff
     "dependencies": [
         "gmp",
 ...
         "boost-python",
         "eigen3",
-       "vcpkg-get-python-packages", {
+       {
            "name": "vcpkg-get-python-packages",
            "host": true
        }
    ]
~~~

The second thing is not easy.
> I think I also need to add vcpkg_add_to_path to my CMakeLists.txt file, but adding it above my project declaration yields Unknown CMake command, so this isn't right.

What vcpkg-get-python-packages provides is a script function to be used in `portfiles.cmake`, CMake script mode.
But probably you are not writing a port, but a "top-level" project, `CMakeLists.txt`, CMake project mode.
https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide#portfiles-are-run-in-script-mode
Using the dependency provides nothing for a top-level project. I cannot help fill the gap.

**Reply by ofloveandhate** [reply]
> But probably you are not writing a port, but a "top-level" project, CMakeLists.txt, CMake project mode

True.

**Reply by ofloveandhate** [reply]
Thank you for your help.  I'll try another method to get my package compiled and usable on Windows.
